server/debug/src/symbol.rs

```diff
@@ -147,52 +162,127 @@ impl SymbolHandler {
             pSymInfo: winapi::PSYMBOL_INFOW, SymbolSize: winapi::ULONG, UserContext: winapi::PVOID
         ) -> winapi::BOOL
-            where F: FnMut(&Symbol, usize) -> bool
+            where F: FnMut(Symbol, usize) -> bool
         {
             let symbol = &*pSymInfo;
-            let Context {
-                ref symbols,
-                ref mut f,
-            } = *(UserContext as *mut Context<F>);
-
-            let mut module = winapi::IMAGEHLP_MODULEW64 {
-                SizeOfStruct: mem::size_of::<winapi::IMAGEHLP_MODULEW64>() as winapi::DWORD,
-                .. mem::zeroed()
-            };
-            if dbghelp::SymGetModuleInfoW64(
-                symbols.0, symbol.Address, &mut module
-            ) == winapi::FALSE {
-                return winapi::TRUE;
-            }
-
-            let mut tag: winapi::DWORD = 0;
-            if dbghelp::SymGetTypeInfo(
-                symbols.0, module.BaseOfImage, symbol.Index,
-                winapi::TI_GET_SYMTAG, &mut tag as *mut _ as *mut _
-            ) == winapi::FALSE {
-                return winapi::TRUE;
-            }
-
-            if tag != 5 {
-                return winapi::TRUE;
-            }
+            let mut f = &mut *(UserContext as *mut F);
 
             let symbol = Symbol {
                 // for some reason, pSymInfo.NameLen includes the null terminator here
                 name: OsString::from_wide_raw(symbol.Name.as_ptr()),
                 address: symbol.Address as usize,
                 size: SymbolSize as usize,
+                type_index: symbol.TypeIndex as u32,
                 flags: symbol.Flags,
             };
-            if f(&symbol, SymbolSize as usize) { winapi::TRUE } else { winapi::FALSE }
+            if f(symbol, SymbolSize as usize) { winapi::TRUE } else { winapi::FALSE }
+        }
+    }
+
+    pub fn type_from_index(&self, module: usize, type_index: u32) -> io::Result<Type> {
+        let tag: winapi::SymTag = self.get_type_info(module, type_index)?;
+        if tag == winapi::SymTagBaseType {
+            let base = self.get_type_info::<BasicType>(module, type_index)?;
+
+            use self::BasicType::*;
+            let base = match base {
+                NoType | Void => Primitive::Void,
+                Bool => Primitive::Bool,
+                Char | WChar | Int | Long => Primitive::Int { signed: true },
+                UInt | ULong => Primitive::Int { signed: false },
+                Float => Primitive::Float,
+                _ => return Err(io::Error::new(io::ErrorKind::Other, "unsupported type")),
+            };
+
+            let TypeLength(size) = self.get_type_info(module, type_index)?;
+
+            Ok(Type::Base { base: base, size: size as usize })
+        } else if tag == winapi::SymTagPointerType {
+            let TypeIndex(target) = self.get_type_info(module, type_index)?;
+            Ok(Type::Pointer { type_index: target })
+        } else if tag == winapi::SymTagArrayType {
```

Comment:
What did you have to do to get winapi::SymTagArrayType to work?  I can't find any reference in the winapi crate documentation.